packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -278,8 +295,11 @@ export class ConnectionStateHandler {
             this.receivedRemoveMemberEvent(clientId);
         });
 
-        // if we have a clientId from a previous container we need to wait for its leave message
-        if (this.clientId !== undefined && protocol.quorum.getMember(this.clientId) !== undefined) {
+        // if we have a clientId from a previous container having pending ops, we need to wait for its leave message
+        if (this.clientId !== undefined &&
+            this.handler.shouldClientJoinWrite() &&
```

Comment:
@wes-carlson - Will this work "as expected" in the rehydration case?  This function is ultimately implemented in `ConnectionManager`:

```typescript
    public shouldJoinWrite(): boolean {
        // We don't have to wait for ack for topmost NoOps. So subtract those.
        return this.clientSequenceNumberObserved < (this.clientSequenceNumber - this.trailingNoopCount);
    }
```

Reply:
Jatin and I discussed this and added it to match the other place that restarts the timer - If there aren't local un-ack'd ops then we don't need to bother waiting for the previous client's Leave op.  I added some asserts in `receivedConnectEvent` function above and want to keep the state consistent.

Reply:
I'm not sure why this should change. `initProtocol()` is called as part of container load, so I think `shouldJoinWrite()` will always return false because we haven't submitted ops. If we load with stashed ops, we will need to wait for our old client ID, but this check won't work because we didn't submit the ops in this container.

Reply:
In the reconnect flow, we only need to wait if there may be unack'd ops, to avoid duplicating those ops. In the stashed ops case, is there a chance some of the stashed ops have already been seen by the ordering service? Or is there another reason you wait for the Leave op?

Reply:
> In the stashed ops case, is there a chance some of the stashed ops have already been seen by the ordering service?

Yes. Ops are stashed while pending (unACKed) so the same precautions need to be taken.

In `initProtocol()`,  `this.clientId` will be undefined if we don't have stashed ops, or if we have stashed ops from a container that never connected. In these cases, it is safe to move immediately to "connected".

Reply:
Ok. After hydration:
- could the first connection be a read connection, or will it always be write?
- How is the ConnectionManager initialized, in terms of its tracking of client sequence numbers? Can those values (used in calculating shouldJoinWrite) be trusted?

Reply:
> could the first connection be a read connection, or will it always be write?

It should always be write if we have stashed ops I think. I believe we always connect write currently anyway

> How is the ConnectionManager initialized, in terms of its tracking of client sequence numbers? Can those values (used in calculating shouldJoinWrite) be trusted?

They aren't, so it will be 0. It will not be correct for initial connect with stashed ops I think, but it isn't currently used in this scenario. Once we reach "connected" state, the ops will be resubmitted and it should work the same as normal reconnect.

Reply:
Ok. Then I will revert the change in initProtocol but leave the two asserts in receivedConnectEvent. Sound good?